### PR TITLE
fix(canal): Preserve binlog filename received in the fake rotate event

### LINKED ISSUE #406: Canal: Wrong binlog position after MariaDB is restarted
**STR**

1. Start syncing
2. Stop MariaDB
3. Start MariaDB
4. Ensure that canal is reconnected.
5. Commit some transaction in the database.

**Expected**

Binlog position is correct in the OnXID handler.

**Actual**

`pos.Pos` is updated to the new position, but `pos.Name` is wrong.

**Environment**

MariaDB 10.3.14

MariaDB creates a new binlog file on start and sends fake rotate events with a new binlog filename and position. These events are ignored in canal as a consequence of the #397 so `Pos` is updated with LogPos from headers of the received events but `Name` stays the same as before reconnection.

I believe that issue should be reproducible with MySQL but I don't have one to verify this.

**Logs**
```
[info] sync.go:57 MariadbGTIDEvent: 1841
[info] sync.go:66 CurPos: (mysql-bin.000017, 1841)
[info] sync.go:57 QueryEvent: 2201
[info] sync.go:66 CurPos: (mysql-bin.000017, 2201)
[info] binlogsyncer.go:698 receive EOF packet, retry ReadPacket
[error] binlogsyncer.go:639 connection was bad
[info] binlogsyncer.go:581 begin to re-sync from (mysql-bin.000017, 2201)
...
[info] binlogsyncer.go:211 register slave for master server mysql:3306
[error] binlogsyncer.go:239 kill connection 17 error ERROR 1094 (HY000): Unknown thread id: 17
[info] binlogsyncer.go:245 kill last connection id 17
[info] binlogsyncer.go:731 rotate to (mysql-bin.000017, 2201)
[info] binlogsyncer.go:731 rotate to (mysql-bin.000018, 4)
[info] sync.go:57 RotateEvent: 0
[info] sync.go:57 FormatDescriptionEvent: 0
[info] sync.go:57 StopEvent: 2224
[info] sync.go:66 CurPos: (mysql-bin.000017, 2224)
[info] sync.go:57 RotateEvent: 0
[info] sync.go:57 FormatDescriptionEvent: 256
[info] sync.go:66 CurPos: (mysql-bin.000017, 256)
[info] sync.go:57 MariadbGTIDListEvent: 299
[info] sync.go:66 CurPos: (mysql-bin.000017, 299)
[info] sync.go:57 MariadbBinLogCheckPointEvent: 342
[info] sync.go:66 CurPos: (mysql-bin.000017, 342)
[info] sync.go:57 MariadbGTIDEvent: 384
[info] sync.go:66 CurPos: (mysql-bin.000017, 384)
[info] sync.go:57 TableMapEvent: 502
[info] sync.go:66 CurPos: (mysql-bin.000017, 502)
[info] sync.go:57 DeleteRowsEventV1: 616
[info] sync.go:66 CurPos: (mysql-bin.000017, 616)
[info] sync.go:57 XIDEvent: 647
[info] sync.go:66 CurPos: (mysql-bin.000017, 647)
```

CurPos must be (mysql-bin.000018, 647) for XIDEvent.

Events and log pos are written by `log.Infof("%v: %v", ev.Header.EventType, ev.Header.LogPos)` before checking for header's `LogPos` and `log.Infof("CurPos: %v", pos)` before the switch.

To be honest I'm not sure how ignoring fake rotate events helps to mitigate issue in #396 
Probably @csuzhangxc or @lintanghui can shed some light on this.

### PR DESCRIPTION
**Pull request by mefcorvi**
The name of a binlog file received in the fake rotate event must be preserved as it was discussed in #406. 

The decision is based on comments in 
https://github.com/mysql/mysql-server/blob/8e797a5d6eb3a87f16498edcb7261a75897babae/sql/rpl_binlog_sender.h#L235

```
... It contains the name of the binlog we are going to
    send to the slave.
    Faked rotate event is required in a few cases, so slave can know which
    binlog the following events are from.
  - The binlog file slave requested is Empty. E.g.
    "CHANGE MASTER TO MASTER_LOG_FILE='', MASTER_LOG_POS=4", etc.
  - The position slave requested is exactly the end of a binlog file.
  - Previous binlog file does not include a rotate event.
    It happens when server is shutdown and restarted.
  - The previous binary log was GTID-free (does not contain a
    Previous_gtids_log_event) and the slave is connecting using
    the GTID protocol.
```

Closes #406

### CODE REVIEW

**Comment by siddontang** [comment]
thanks @mefcorvi 

can we add a test to cover this?

**Comment by mefcorvi** [comment]
> can we add a test to cover this?

I think I can cover the first case "The binlog file slave requested is Empty", but for other cases I need to be able to manage the mysql server from the tests, i.e. shut it down, restart, reconfigure. Can you advise the simplest way to do that?

File: canal/sync.go [hunk 1]
```diff
@@ -51,19 +56,35 @@ func (c *Canal) runSyncBinlog() error {
...
 		force = false
 		pos := c.master.Position()
 
 		curPos := pos.Pos
+
+		// If log pos equals zero then the received event is a fake rotate event and
+		// contains only a name of the next binlog file
+		// See https://github.com/mysql/mysql-server/blob/8e797a5d6eb3a87f16498edcb7261a75897babae/sql/rpl_binlog_sender.h#L235
+		// and https://github.com/mysql/mysql-server/blob/8cc757da3d87bf4a1f07dcfb2d3c96fed3806870/sql/rpl_binlog_sender.cc#L899
+		if ev.Header.LogPos == 0 {
```
Comment (siddontang) [inline canal/sync.go]: why move the check from above to here?

Comment (mefcorvi) [reply canal/sync.go]: That's a good question. I've restored its original position.

**Pull request merged**